### DISCUSSION TITLE
ModuleNotFoundError: No module named 'numpy'

### DISCUSSION THREAD

**Discussion by caelrw** [SiEPIC/SiEPIC_EBeam_PDK#311] [Q&A]
On Ubuntu 24.04, I installed this package in to KLayout 0.29.8 using Python_packages_for_KLayout 1.2, siepic_tools 0.5.14 and siepic_ebeam_pdk 0.4.27. Attempting to open the MZI example layout returns the following error:

ModuleNotFoundError: No module named 'numpy'

I tried:
1. Switching from Python 3.12 to Python 3.11 in an anaconda environment
2. Deleting and re-installing the packages from KLayout
3. On my end, numpy is already installed.

**Comment by lukasc-ubc** [comment]
Are you running the example within KLayout Application?  If so, please try this:

Open the Macro Editor (F5).  In the Console, choose Python, and type in:

> from SiEPIC.install import install
> install ('numpy')

**Comment by caelrw** [comment]
I found the issue was with pip being called from **siepic_ebeam_pdk 0.4.27 - SiEPIC EBeam PDK**
I  deactivated the conda environment and worked outside anaconda. After using the macro editor and the lines

```
from SiEPIC.install import install
install('numpy')
```

The error box returned that 'pip' was missing, and that despite declaring a pyenv version 3.10 I was still using the system python 3.12.3 such that pip was not supported. I couldn't figure out how to make klayout recognize pyenv python for the tool packages. **The fix for me was instead to enable pip and then install the dependent libraries using the following commands**:

`python -m ensurepip --default-pip`

which I found instructions for at this website: https://codedamn.com/news/python/how-to-fix-no-module-named-pip

When pip was enabled, consecutive libraries 'numpy', 'scipy' were then indicated as also not being installed returning 
```
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.

```
I installed these packages with
```
sudo apt install python3-numpy
 sudo apt install python3-scipy

```
In this process, I received the following message:
```
WARNING: pip is being invoked by an old script wrapper. This will fail in a future version of pip.
Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.

```
Therefore, it seems that the pip command is out-dated, which is why the package manager cannot work with newer versions of python. It may be that pip won't be available in the future, or a slightly different command will be required.

**Comment by lukasc-ubc** [comment]
Can you please clarify:  are you running KLayout as a GUI Application?  Or are you running it in Python (VSCode, etc), and using “import klayout, siepic_ebeam_pdk …”?

From what you write below, it looks like it is within the GUI, but with an external Python environment.  In this case, you should be able to install the packages externally, as long as KLayout is built to the corresponding python installation.

I am more familiar with the Mac installation, which comes in “heavyweight” and “lightweight”. The hw version includes a full python environment, and the only way to install packages is using the deprecated pip function.  For the lw version, it is similar to the Linux one, where you can install packages externally as you are doing.

So you have numpy working now?

**Comment by caelrw** [comment]
I am running KLayout as a GUI application, but on Ubuntu 24.04. On this Linux OS, KLayout calls the system python version which is Python 3.12.3. However, Python 3.12.3 does not have pip but pip3 by default, and installing earlier Python version  affects the Linux OS package manager if not installed and used separately. Instead, I brought back pip, then numpy and scipy with Python 3.12.3 so I can successfully use the SiEPIC packages in KLayout with Python 3.12.3 with no errors. Python 3.12.3 can use the deprecated pip, but future Python versions may not.

I have pip working now so the SiEPIC packages can import updated pip, scipy, and numpy to KLayout.

**Comment by lukasc-ubc** [comment]
Could you not install numpy using pip3?  Numpy is such a standard package that all Python installations should easily have it.

**Reply by caelrw** [reply]
Yes, I had installed numpy with pip3, but the deprecated pip command from KLayout would not recognize it and still flag errors with the SiEPIC libraries without the extra command

`python -m ensurepip --default-pip`

from the terminal.

**Comment by MatteoPenna** [comment]
I was having the same issues after updating to KLayout version 0.28.17. The problem stemmed from (a) the pip command used by SiEPIC.install to install the correct Python dependencies was not working, and (b) the "heavyweight" (HW) version of KLayout I had installed will create its own Python environment with a specific version of Python (3.11.7 in my case) which is not directly in my python environment path, so typing "pip3 install numpy" in my terminal is no help. 

The fix I found was as follows:

1. Open KLayout Macro editor.

3. In the Python console type: 
	> import sys
	> print(sys.version)
	> print(sys.path)

4. The "print(sys.path)" returns the directory of the Python installation that KLayout uses. With this directory, we now know where to install numpy, scipy, and whatever other dependencies are required. Dependencies should be located in the /lib/pythonx.xx folder of the returned directory, ../lib/python3.11 in my case. For example, in my case "print(sys.path)" returned:
	> "/Applications/klayout 28.app/Contents/Frameworks/Python.framework/Versions/3.11/lib/python3.11" 
5. Open a terminal window. Navigate to parent directory the returned directory, i.e., the lib directory. In my case:
	> "/Applications/klayout 28.app/Contents/Frameworks/Python.framework/Versions/3.11/lib/"
6. Now install the required dependencies into the python3.xx directory. Run in the terminal:
	> pip install --target ./python3.xx/ package1 package2 package3 etc 

    In my case I ran:
    > pip install --target ./python3.11/ numpy scipy requests pyparsing packaging 

After this I had no issues running any SiEPIC PCells or SiEPIC tools functionality. My KLayout also stopped prompting me every time I opened it to install those packages... which makes sense. Hope this helps someone!

**Comment by lukasc-ubc** [comment]
Is there a reason you are using an old version (0.28) of KLayout, rather than the latest 0.29?

**Reply by MatteoPenna** [reply]
I tried KLayout 29 and I had the same original error. I suspect that this fix for 28 would also work for 29.
